networking_cisco/plugins/cisco/device_manager/plugging_drivers/aci_vlan_trunking_driver.py

```diff
@@ -0,0 +1,255 @@
+        self._get_ext_net_name = None
+        self._default_ext_dict = DEFAULT_EXT_DICT
+        self._transit_nets_cfg = {}
+
+    @property
+    def transit_nets_cfg(self):
+        if self._cfg_file:
+            networks_dict = open(self._cfg_file, 'r').read()
+            self._transit_nets_cfg = eval(networks_dict)
+        else:
+            self._transit_nets_cfg = {}
+        return self._transit_nets_cfg
+
+    @property
+    def get_ext_net_name(self):
+        # FIXME(tbachman): sadly, only exists to enable UT
+        if self.apic_driver:
+            return self._get_ext_net_name
+        else:
+            return None
+
+    @property
+    def get_vrf_context(self):
+        if self.apic_driver:
+            return self._get_vrf_context
+        else:
+            return None
+
+    def _get_vrf_context_gbp(self, context, router_id, port_db):
+        ptg, pt = self.apic_driver._port_id_to_ptg(context, port_db['id'])
+        #l3p = _get_ptg_l3p(context, ptg)
+
+    def _get_vrf_context_ml2(self, context, router_id, port_db):
+        router = self.l3_plugin.get_router(context, router_id)
+        return {'vrf_id': router['tenant_id']}
+        #return {'vrf_id': port_db['tenant_id']}
+
+    def _get_external_network_dict(self, context, port_db):
+        """Get external network information
+
+        Get the information about the external network,
+        so that it can be used to create the hidden port,
+        subnet, and network.
+        """
+        network = self._core_plugin.get_network(context,
+            port_db.network_id)
+
+        # network names in GBP workflow need to be reduced, since
+        # the network may contain UUIDs
+        external_network = self.get_ext_net_name(network['name'])
+        return self.transit_nets_cfg.get(
+            external_network) or self._default_ext_dict
+
+    @property
+    def l3_plugin(self):
+        if not self._l3_plugin:
+            self._l3_plugin = manager.NeutronManager.get_service_plugins().get(
+                svc_constants.L3_ROUTER_NAT)
+        return self._l3_plugin
+
+    @property
+    def apic_driver(self):
+        """Get APIC driver
+
+        There are different drivers for the GBP workflow
+        and Neutron workflow for APIC. First see if the GBP
+        workflow is active, and if so get the APIC driver for it.
+        If the GBP service isn't installed, try to get the driver
+        from the Neutron (APIC ML2) workflow.
+        """
+        if not self._apic_driver:
+            try:
+                self._apic_driver = (
+                    manager.NeutronManager.get_service_plugins()[
+                        'GROUP_POLICY'].policy_driver_manager.policy_drivers[
+                            'apic'].obj)
+                self._get_vrf_context = self._get_vrf_context_gbp
+                self._get_ext_net_name = self._get_ext_net_name_gbp
+            except Exception:
+                pass
+            if not self._apic_driver:
```

Comment:
I don't think it is -- if the GBP driver gets assigned above, we don't want to try loading the ML2 driver.